### PR TITLE
fix: allow updates on a sharded collection COMPASS-6058

### PR DESCRIPTION
**Pull request by alenakhineika**
## Description

The `findAndModify` command is not allowed on sharded collections and throws the `"Query for sharded findAndModify must contain the shard key"` error. We check if the command returned the `ShardKeyNotFound` error name, we try using updateOne/replaceOne instead of findOneAndUpdate/findOneAndReplace to perform the update operation.

### Checklist
- [x] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [x] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [x] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-crud/src/stores/crud-store.ts [hunk 1]
```diff
@@ -1584,44 +1570,104 @@ type ErrorOrResult =
 export async function findAndModifyWithFLEFallback(
   ds: DataService,
   ns: string,
-  doFindAndModify: (
-    ds: DataService,
-    ns: string,
-    opts: { returnDocument: 'before' | 'after'; promoteValues: false }
-  ) => Promise<ErrorOrResult>
+  query: Partial<QueryState>,
+  object: unknown,
```
Comment (addaleax) [inline packages/compass-crud/src/stores/crud-store.ts:1574]: These types don't seem quite right. `query` is not related to `queryState`, and `object` is typed as `unknown` here but we always cast it using `as Document` below

File: packages/compass-crud/src/stores/crud-store.ts [hunk 2]
```diff
@@ -1584,44 +1570,104 @@ type ErrorOrResult =
...
+          options
+        );
       } else {
-        [d] = docs;
+        document = await ds.findOneAndReplace(
+          ns,
+          query,
+          object as Document,
+          options
+        );
```
Comment (addaleax) [inline packages/compass-crud/src/stores/crud-store.ts:1649]: There's quite a bit of duplication now, this code block shows up 4 times in the same function essentially... I'd either store the arguments as an array and then simplify these calls to `await ds.findOneAndReplace(...args);`, or dynamically select the method (`const method = modificationType === 'update' ? 'findOneAndUpdate' : 'findOneAndReplace'; await ds[method](ns, query, object, options);`)

File: packages/data-service/src/data-service.ts [hunk 3]
```diff
@@ -1488,6 +1520,28 @@ class DataServiceImpl extends WithLogContext implements DataService {
     return (await coll.findOneAndUpdate(filter, update, options)).value;
   }
 
+  @op(mongoLogId(1_001_000_183))
```
Comment (addaleax) [inline packages/data-service/src/data-service.ts:1523]: Did the script suggest this number? It should always be suggesting one that's higher than all other used numbers

Comment (alenakhineika) [reply packages/data-service/src/data-service.ts:1523]: It was, I think the problem is that I merged the main a couple of times and for this actually, it was fine, but for replaceOne 1_001_000_184 was errored as a duplicate and I keep changing it.

**Review by addaleax — APPROVED** [review]

**Pull request merged**